tests/parametric/test_dynamic_configuration.py

```diff
@@ -0,0 +1,226 @@
+"""
+Test the dynamic configuration via Remote Config (RC) feature of the APM libraries.
+"""
+import json
+import random
+from typing import Any
+from typing import Dict
+
+from utils import missing_feature, context, scenarios
+
+import pytest
+
+
+parametrize = pytest.mark.parametrize
+
+
+base_env = {
+    "DD_TELEMETRY_HEARTBEAT_INTERVAL": "0.2",
+    "DD_REMOTE_CONFIG_POLL_INTERVAL_SECONDS": "0.2",
+}
+
+DEFAULT_SAMPLE_RATE = 1.0
+
+
+def send_and_wait_trace(test_library, test_agent, **span_kwargs):
+    with test_library.start_span(**span_kwargs):
+        pass
+    traces = test_agent.wait_for_num_traces(num=1, clear=True)
+    assert len(traces) == 1
+    return traces[0]
+
+
+def set_and_wait_rc(test_agent, config: Dict[str, Any]):
+    cfg = {
+        # v1 dynamic config
+        "tracing_sampling_rate": None,
+        "log_injection_enabled": None,
+        "tracing_header_tags": None,
+        # v2 dynamic config
+        "runtime_metrics_enabled": None,
+        "tracing_debug": None,
+        "tracing_service_mapping": None,
+        "tracing_sampling_rules": None,
+        "span_sampling_rules": None,
+        "data_streams_enabled": None,
+    }
+    for k, v in config.items():
+        cfg[k] = v
+
+    cfg_id = "%032x" % random.getrandbits(128)
+    test_agent.set_remote_config(
+        path="datadog/2/APM_TRACING/%s/config" % cfg_id,
+        payload={
+            # These values don't matter, can be anything
+            "action": "enable",
+            "service_target": {"service": "myservice", "env": "dev"},
+            "lib_config": cfg,
+        },
+    )
+    test_agent.wait_for_telemetry_event("app-client-configuration-change", clear=True)
+    return test_agent.wait_for_apply_status("APM_TRACING", clear=True, state=2)
+
+
+def assert_sampling_rate(span: Dict, rate: float):
+    """Asserts that a span returned from the test agent is consistent with the given sample rate.
+
+    It is assumed that the span is the root span of the trace.
+    """
+    # TODO: find the right heuristics to assert the sample rate
```

Comment:
If the tracer sends all spans to the Agent, then the first span of each chunk will contain the [_dd.rule_psr][1] "metric" (numeric tag), assuming the sampling decision was made by that tracer due to a sampling rule or `DD_TRACE_SAMPLE_RATE`, which is like a catch-all sampling rule.

If the tracer does _not_ send all spans to the Agent, then your job is harder. I don't know whether any tracers have this behavior. The point of the "tracer stats" project was to calculate `trace.*` metrics in the tracers, instead of in the Agent, so that eventually tracers would not have to send dropped traces to the Agent at all. I don't know how far along we are with that, though.

You could pick a high sample rate, say 97%, and then create 20 or 30 traces and assert that for each you received (you won't necessarily receive them all), the first span of each chunk had the desired `_dd.rule_psr` metric (numeric tag).

By picking a high sample rate, you avoid the test becoming flaky, at least with high probability :)

Ah, it depends on how trace IDs are chosen, too.

[1]: https://github.com/DataDog/dd-trace-cpp/blob/0536a1653c6d689ab5f29d9b9defb6532e4b1b5b/src/datadog/tags.cpp#L25

Reply:
:+1: it seems then that relying solely on `_dd.rule_psr` should be sufficient.

Right now we can assume all traces are sent to the agent. I'll make note of the suggestion for the day when that is not the case in the docstring of this function.